src/language/__tests__/printString-test.js

```diff
@@ -0,0 +1,54 @@
+import { expect } from 'chai';
+import { describe, it } from 'mocha';
+
+import { printString } from '../printString';
+
+describe('printString', () => {
+  it('prints a simple string', () => {
+    expect(printString('hello world')).to.equal('"hello world"');
+  });
+
+  it('escapes quotes', () => {
+    expect(printString('"hello world"')).to.equal('"\\"hello world\\""');
+  });
+
+  it('does not escape single quote', () => {
+    expect(printString("who's test")).to.equal('"who\'s test"');
+  });
+
+  it('escapes backslashes', () => {
+    expect(printString('escape: \\n')).to.equal('"escape: \\\\n"');
+  });
+
+  it('escapes well-known control chars', () => {
+    expect(printString('\b\f\n\r\t')).to.equal('"\\b\\f\\n\\r\\t"');
+  });
+
+  it('escapes zero byte', () => {
+    expect(printString('\x00')).to.equal('"\\u0000"');
+  });
+
+  it('does not escape space', () => {
+    expect(printString(' ')).to.equal('" "');
+  });
+
+  it('escapes all other control chars', () => {
+    for (let i = 1; i <= 0x9f; i++) {
+      const source = String.fromCharCode(i);
+      if (/[\b\f\n\r\t]/.test(source) || (i >= 0x0020 && i <= 0x007e)) {
+        continue;
+      }
+      expect(printString(source)).to.equal(
+        `"\\u00${i <= 0x000f ? '0' : ''}${i.toString(16).toUpperCase()}"`,
+      );
+    }
+  });
+
+  it('does not escape non-ascii character', () => {
+    expect(printString('\u21BB')).to.equal('"\u21BB"');
+  });
+
+  it('does not escape supplementary character', () => {
```

Comment:
This is tricky detail where a GraphQL String literal is not compatible with a JSON string.
I could imagine that the default JSON serialization doesn't handle that correctly.

You could escape it as surrogate pair and make it compatible without any additional work.

Reply:
These are all just characters that a JSON string would encode/escape as it defines.

For example the printed GraphQL string literal `"\u{12345}"` would JSON stringify as `"\"\\u{12345}\""`

Do you have a test case in mind that would cover your concern?

Reply:
JSON also supports all supplemental Unicode code points, escaped or unescaped

Reply:
There should be no case where a GraphQL string literal is directly provided to a JSON parser

Reply:
unescaped unicode code points are fine as you said, I am just thinking about the braced escape as this is something not supported by JSON.

I was just thinking if these printed string are every serialized directly.

For example a Field description like  "This is a funny field \u{1F921}".  (1F921 is = 🤡)

If this GraphQL string would be serialized to a JSON String  it would be printed as 
"This is a funny field \\u{12345}" which is not a correct JSON serialization. The correct JSON serialization would have to use surrogate pair.

If these printed String are never directly serialized, then it should be fine.

Reply:
That's right, a printed GraphQL string literal should be and is never represented directly as parseable JSON.

The one case where we would serialize a GraphQL string literal is during [introspection when returning default values](https://spec.graphql.org/draft/#sel-GAJXVHHDAAACDB78R). But this is always a printed string, then serialized as JSON, so `defaultValue: "123"` is the integer value `123` and `defaultValue: "\"123\""` is the string value `"123"`.

Reply:
ok ...so the default value "\u{1F921}"  would be serialized as "\"\\u{1F921}\"" which is a valid GraphQL String value encoded as JSON. My point was that this is not a direct representation of this code point in JSON: "\uD83E\uDD21" would be a better JSON encoded value, but if you print this value as "\uD83E\uDD21" and then serialize it would be "\"\\uD83E\\uDD21\"\" which is not really better.

Reply:
Agreed. Right now before this PR in the state where we accidentally implicitly support full Unicode and co-opt JSON.stringify to produce string literal grammars, we would render the string value [U+1F921] as the GraphQL string literal `"\uD83E\uDD21"`, which then serialized in a JSON body in an introspection query: "\"\\uD83E\\uDD21\"\"

After this diff with this manual version of string literal printing, we would render that same string value as the GraphQL string literal `"🤡"` which then serialized in a JSON body in an introspection query: "\"\uD83E\uDD21\"\".

IMHO this latter state is preferable. I just pushed a new version which explicitly includes this example in `introspection-test.js` to ensure it behaves as expected

Reply:
sounds good